scope/test.h

```diff
@@ -55,6 +51,37 @@ namespace scope {
     }
   }
 
+  // eval_equal_impl for anything having std::begin and std::end overloads
+  template <typename ExceptionType, typename ExpSequenceT, typename ActSequenceT>
+  auto eval_equal_impl(ExpSequenceT&& e, ActSequenceT&& a, int, const char* const file, int line, const char* msg = "") -> decltype(std::begin(e), std::end(e), std::begin(a), std::end(a), void()) {
+    const auto abeg = std::begin(a);
+    const auto aend = std::end(a);
+    const auto ebeg = std::begin(e);
+    const auto eend = std::end(e);
+    const size_t elen = eend - ebeg;
+    const size_t alen = aend - abeg;
+
+    if (alen != elen) {
```

Comment:
so, here's something I've wondered, is it actually useful to throw an exception as soon as you realize that the lengths are off? might it not be better/more informative to defer until the point of divergence?